sw/device/silicon_creator/rom/e2e/rom_e2e_flash_ctrl_init_test.c

```diff
@@ -0,0 +1,304 @@
+
+  // Check permission bits.
+  //
+  // Expected values:
+  // * EN        = true (otherwise the configuration does not apply)
+  // * RD_EN     = false
+  // * PROG_EN   = false
+  // * ERASE_EN  = false
+  uint32_t info_page_cfg =
+      abs_mmio_read32(kBase + FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_REG_OFFSET);
+  // TODO: the below fails because enable is false, but it's unclear from the
+  // flash documentation if this actually disables all the other non-enabled
+  // things. Need to investigate.
+  // CHECK(kMultiBitBool4True ==
+  //      bitfield_field32_read(info_page_cfg,
+  //                            FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_EN_1_FIELD));
+  CHECK(kMultiBitBool4False ==
+        bitfield_field32_read(info_page_cfg,
+                              FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_RD_EN_1_FIELD));
+  CHECK(kMultiBitBool4False ==
+        bitfield_field32_read(
+            info_page_cfg, FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_PROG_EN_1_FIELD));
+  CHECK(kMultiBitBool4False ==
+        bitfield_field32_read(
+            info_page_cfg, FLASH_CTRL_BANK0_INFO0_PAGE_CFG_1_ERASE_EN_1_FIELD));
+
+  // Check that the info page configuration is write-disabled.
+  uint32_t info_page_regwen =
+      abs_mmio_read32(kBase + FLASH_CTRL_BANK0_INFO0_REGWEN_1_REG_OFFSET);
+  CHECK(bitfield_bit32_read(info_page_regwen,
+                            FLASH_CTRL_BANK0_INFO0_REGWEN_1_REGION_1_BIT) ==
+        false);
+
+  // Compute the page address from the bank and page number (bank 0, page 1).
+  uint32_t data_addr = (0 * FLASH_CTRL_PARAM_BYTES_PER_BANK) +
+                       (1 * FLASH_CTRL_PARAM_BYTES_PER_BANK);
+
+  // Attempt to read, expect an error.
+  uint32_t read_data[1];
+  CHECK(flash_ctrl_data_read(data_addr, 1, read_data) ==
+        kErrorFlashCtrlDataRead);
+
+  // Attempt to write, expect an error.
+  uint32_t write_data[1] = {0};
+  CHECK(flash_ctrl_data_write(data_addr, 1, write_data) ==
+        kErrorFlashCtrlDataWrite);
+
+  // Attempt to erase, expect an error.
+  CHECK(flash_ctrl_data_erase(data_addr, kFlashCtrlEraseTypePage) ==
+        kErrorFlashCtrlDataErase);
+}
+
+/**
+ * Verify that the flash controller is initialized.
+ */
+static void is_initialized_test(void) {
+  // Check that the physical status register does not say flash controller is
+  // still initializing.
+  uint32_t phy_status =
+      abs_mmio_read32(kBase + FLASH_CTRL_PHY_STATUS_REG_OFFSET);
+  CHECK(bitfield_bit32_read(phy_status, FLASH_CTRL_PHY_STATUS_INIT_WIP_BIT) ==
+        false);
+
+  // Check that the general status register does not say flash controller is
+  // still initializing, and that the FIFOs are empty.
+  uint32_t status = abs_mmio_read32(kBase + FLASH_CTRL_STATUS_REG_OFFSET);
+  CHECK(bitfield_bit32_read(status, FLASH_CTRL_STATUS_INIT_WIP_BIT) == false);
+  CHECK(bitfield_bit32_read(status, FLASH_CTRL_STATUS_RD_EMPTY_BIT) == true);
+  CHECK(bitfield_bit32_read(status, FLASH_CTRL_STATUS_PROG_EMPTY_BIT) == true);
```

Comment:
The `init_wip` bits were the best way I could determine from the docs to check if flash was initialized. However, these are false at reset, so if some failure caused the initialization to never start then this test wouldn't catch it. I'd be interested to know if there's a better way to check this.

Reply:
hmm... that's good feedback, I probably should have had an actual status that just indicated whether we've initialized. 
I can add that if we think it helps? 

Without that, the best I could suggest is that when you initialize, first poll for wip to become high, then check for it to be false later.

Reply:
Would the following work?
```
(pseudo-code)
Verify that flash_ctrl.INIT.VAL == 1 // rw1s, requested init.
Verify that INIT_WIP == 0 // ro, 0 if no need to panic
```

Reply:
This works, perfect!